### PR TITLE
Custom django cli cmd to recompute quality score

### LINKED ISSUE #1183: Recalculate quality score for collections that have a score
### Use Case
The quality score is calculated at the time of import, edits to scoring bits won’t update the quality score until the collection is re-imported. The types of edits are:
1. Severity change for an existing rule (or to stop counting a rule towards the score)
2. Weight change for how a particular severity impacts the score 
3. New or edited ansible-lint rules (or yamllint config)

### Proposed Solution for Severity and Weight Changes
- Use custom django-admin command to recompute the quality score for the collection and its content
- Only interact with collections which already have a quality score
- Use existing linter violations to feed into current/edited rule_id to severity mapping and any weight edits
- Do not reimport or relint the collection
- Allow cli to be run on collections scored before or after a certain date
- Update collection’s quality score timestamp

### Note for New and Edited Rules
For a new or edited ansible-lint rule, the only way to rescore will be for the author to reimport. We will not automatically reimport / relint because:
- Github api rate limits for pulling down all scored repos in batch
- The github repo may have changed since author imported - it may not be importable or even exist
- Auto reimporting (or partial reimporting in order to relint) would update galaxy metadata and gh release versions which author may not expect
- Auto reimporting would trigger unexpected notifications to collection followers

Future solution thoughts:
- After pulp is in place, attempt a solution for new and edited rules
- With pulp the artifact will be static and won’t have changed - so relinting should have less potential problems, and won't need to interact with github api
- Consider caching all repos as they are imported - mazer install them

### PR DESCRIPTION
**Pull request by awcrosby**
Closes #1183

### CODE REVIEW

File: galaxy/main/management/commands/recompute_quality.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        sys.stdout = self.stdout
+        self.file.close()
+
+
+class Command(BaseCommand):
+    help = ('Recomputes quality score for collections which already have a '
+            'quality score. Does not reimport or relint - '
+            'uses existing linter violations.')
+    repos = []
+    scored_all, scored_before, scored_after = None, None, None
```
Comment (cutwater) [inline galaxy/main/management/commands/recompute_quality.py]: Please don't use expansion syntax for initializing variables.

File: galaxy/main/management/commands/recompute_quality.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        parser.add_argument(
+            '--scored-after', nargs=1, type=str,
+            help=('Looks at collections scored after given datetime, '
+                  'use format 2018-10-29-23:30:00')
+        )
+
+    def handle(self, *args, **kwargs):
+        try:
+            self.validate_args(kwargs)
+        except Exception as e:
```
Comment (cutwater) [inline galaxy/main/management/commands/recompute_quality.py]: Please use specific exception or multiple exceptions if possible

File: galaxy/main/management/commands/recompute_quality.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            self.count()
+        elif kwargs['recompute']:
+            if not self.repos.count():
+                return self.style.WARNING('0 collections to recompute')
+            self.recompute()
+
+        return self.style.SUCCESS('Successfully completed command')
+
+    def validate_args(self, k):
+        actions = map(lambda x: (x is not None) and (x is not False),
```
Comment (cutwater) [inline galaxy/main/management/commands/recompute_quality.py]: `map(bool, ...)`

File: galaxy/main/management/commands/recompute_quality.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                      [k['count'], k['recompute']])
+        if sum(actions) != 1:
+            raise Exception('Must have one and only one the following args: '
+                            '--count --recompute')
+
+        scored = map(lambda x: (x is not None) and (x is not False),
```
Comment (cutwater) [inline galaxy/main/management/commands/recompute_quality.py]: ditto

File: galaxy/main/management/commands/recompute_quality.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                     [k['scored_all'], k['scored_before'], k['scored_after']])
+        if sum(scored) != 1:
+            raise Exception('Must have one and only one --scored-* arg')
```
Comment (cutwater) [inline galaxy/main/management/commands/recompute_quality.py]: Please do not raise base `Exception` class

File: galaxy/worker/tasks.py [hunk 6]
```diff
@@ -317,20 +317,21 @@ def _update_quality_score(import_task):
         content.save()
         repo_points += content.quality_score
 
-        LOG.debug('content - ids, weights, score: {}, {}, {}'.format(
+        # print vs LOG.debug has cleaner celery + recompute_quality log output
+        print('content - ids, weights, score: {}, {}, {}'.format(
```
Comment (cutwater) [inline galaxy/worker/tasks.py]: To be discussed.

File: galaxy/main/management/commands/recompute_quality.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        repo.refresh_from_db()
+
+        msg = ("{:.1f} -> {:.1f} quality score change for import_task: {}, "
+               "collection: {} {}")
+        self.stdout.write(msg.format(old_score,
+                                     repo.quality_score,
+                                     latest_import_task.id,
+                                     repo.provider_namespace.name,
+                                     repo.name,
+                                     ))
```
Comment (cutwater) [inline galaxy/main/management/commands/recompute_quality.py]: Please prefer the following formatting style:

```
self.stdout.write(msg.format(
    old_score,
    repo.quality_score,
    latest_import_task.id,
    repo.provider_namespace.name,
    repo.name,
))

```

File: galaxy/main/management/commands/recompute_quality.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        )
+        parser.add_argument(
+            '--scored-after', nargs=1, type=str,
+            help=('Looks at collections scored after given datetime, '
+                  'use format 2018-10-29-23:30:00')
+        )
+
+    def handle(self, *args, **kwargs):
+        try:
+            self.validate_args(kwargs)
```
Comment (alikins) [inline galaxy/main/management/commands/recompute_quality.py]: Could validate_args be replaced with a argparse mutually exclusive groups, ala https://docs.python.org/3/library/argparse.html#mutual-exclusion ?

File: galaxy/worker/tasks.py (hunk 6 above)
Comment (awcrosby) [reply galaxy/worker/tasks.py]: I switched this back to logger and fixed the underlying issue #1288

File: galaxy/main/management/commands/recompute_quality.py (hunk 8 above)
Comment (awcrosby) [reply galaxy/main/management/commands/recompute_quality.py]: Yes that worked nicely

**Pull request merged**